### PR TITLE
Add affine grid generator op

### PR DESCRIPTION
**Pull request by wanghaoshuang**

### CODE REVIEW

File: paddle/fluid/operators/affine_grid_cudnn_op.cu.cc [hunk 1]
```diff
@@ -0,0 +1,110 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline paddle/fluid/operators/affine_grid_cudnn_op.cu.cc]: 2016 -> 2018

File: paddle/fluid/operators/affine_grid_op.cc [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+REGISTER_OPERATOR(affine_grid, ops::AffineGridOp, ops::AffineGridOpMaker,
+                  ops::AffineGridGradMaker);
+REGISTER_OPERATOR(affine_grid_grad, ops::AffineGridOpGrad);
+
+REGISTER_OP_CPU_KERNEL(
+    affine_grid,
+    ops::AffineGridOpKernel<paddle::platform::CPUDeviceContext, float>);
+REGISTER_OP_CPU_KERNEL(
+    affine_grid_grad,
+    ops::AffineGridGradOpKernel<paddle::platform::CPUDeviceContext, float>);
```
Comment (qingqing01) [inline paddle/fluid/operators/affine_grid_op.cc]: also register double type.

File: paddle/fluid/operators/affine_grid_op.cc [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+class AffineGridOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput(
+        "Theta",
+        "(Tensor) A batch of affine transform parameters with shape [N, 2, 3]. "
+        "It is used to transform coordinate (x_0, y_0) to coordinate (x_1, "
+        "y_1).");
+    AddInput("Size",
```
Comment (qingqing01) [inline paddle/fluid/operators/affine_grid_op.cc]: Size -> OutputShape or Shape ?

File: paddle/fluid/platform/cudnn_helper.h [hunk 4]
```diff
@@ -80,6 +80,10 @@ enum class PoolingMode {
   kMaximumDeterministic,
 };
 
+enum class SamplerType {
```
Comment (qingqing01) [inline paddle/fluid/platform/cudnn_helper.h]: It seems SamplerType is not used.

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -5756,6 +5757,124 @@ def crop(x, shape=None, offsets=None, name=None):
     return out
 
 
+def affine_grid_gen(theta, out_shape):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Maybe `affine_grid_generator` is betther ?  also need to add `name` in the API.

Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: Maybe `affine_grid_generator` is betther ?

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -5756,6 +5757,124 @@ def crop(x, shape=None, offsets=None, name=None):
...
+                    [-0.5  1.   1. ]
+                    [ 0.   1.   1. ]
+                    [ 0.5  1.   1. ]
+                    [ 1.   1.   1. ]]
+          Step3:
+              Compute output by equation $$Output[i] = C_ * Theta[i]^T$$
+
+    Args:
+        theta (Variable): A batch of affine transform parameters with shape [N, 2, 3].
+        size (Variable | list | tuple): The target output image shape with format [N, C, H, W].
```
Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: `out_shape` is used in above interafce.
size -> out_shape

Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: `out_shape` is used in above interafce.
size -> out_shape

The target output image shape with format [N, C, H, W]. out_shape can be a Variable or a list or tuple.

Comment (qingqing01) [inline python/paddle/fluid/layers/nn.py]: `out_shape` is used in above interafce.
size -> out_shape

File: python/paddle/fluid/tests/unittests/op_test.py [hunk 7]
```diff
@@ -361,8 +361,8 @@ def _get_places(self):
             else:
                 return []
         places = [fluid.CPUPlace()]
-        if core.is_compiled_with_cuda() and core.op_support_gpu(self.op_type):
-            places.append(core.CUDAPlace(0))
+        #        if core.is_compiled_with_cuda() and core.op_support_gpu(self.op_type):
+        #            places.append(core.CUDAPlace(0))
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/op_test.py]: need to resume these code.

**Review by qingqing01 — APPROVED** [review]

**Review by qingqing01 — DISMISSED** [review]

**Review by qingqing01 — DISMISSED** [review]

File: python/paddle/fluid/tests/unittests/test_layers.py [hunk 8]
```diff
@@ -865,6 +865,22 @@ def iou_similarity(self):
...
+
+            theta = layers.data(name="theta", shape=[2, 3], dtype="float32")
+            out_shape = layers.data(
+                name="out_shape", shape=[-1], dtype="float32")
+            data_0 = layers.affine_grid_generator(theta, out_shape)
+            data_1 = layers.affine_grid_generator(theta, [5, 3, 28, 28])
+
+            self.assertIsNotNone(data_0)
+            self.assertIsNotNone(data_1)
+        print(str(program))
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_layers.py:882]: delete?

File: python/paddle/fluid/layers/nn.py [hunk 9]
```diff
@@ -6072,6 +6073,127 @@ def crop(x, shape=None, offsets=None, name=None):
...
 
+def affine_grid_generator(theta, out_shape, name=None):
+    """
+    It generates a grid of (x,y) coordinates using the parameters of
+    the affine transformation that correspond to a set of points where
+    the input feature map should be sampled to produce the transformed
+    output feature map.
+
+
+    Crop input into output, as specified by offsets and shape.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: what is `offset`?

File: python/paddle/fluid/layers/nn.py [hunk 10]
```diff
@@ -6072,6 +6073,127 @@ def crop(x, shape=None, offsets=None, name=None):
...
+              theta = [[[x_11, x_12, x_13]
+                        [x_14, x_15, x_16]]
+                       [[x_21, x_22, x_23]
+                        [x_24, x_25, x_26]]]
+      
+              out_shape = [2, 3, 5, 5]
+      
+          Step 1:
+      
+              Generate relative coordinates according to Size.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: what is `Size`?

File: python/paddle/fluid/layers/nn.py [hunk 11]
```diff
@@ -6072,6 +6073,127 @@ def crop(x, shape=None, offsets=None, name=None):
...
+              The values of relative coordinates are in the interval between -1 and 1.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: Since I don't know what is `size`, I don't what is `relative coordinates` either...

**Review by panyx0718 — COMMENTED** [review]
Need to have better comments

File: python/paddle/fluid/layers/nn.py [hunk 12]
```diff
@@ -6072,6 +6073,124 @@ def crop(x, shape=None, offsets=None, name=None):
     return out
 
 
+def affine_grid_generator(theta, out_shape, name=None):
```
Comment (heavengate) [inline python/paddle/fluid/layers/nn.py]: `@templatedoc()`
`def affine_grid_generator(theta, out_shape, name=None):`

File: python/paddle/fluid/layers/nn.py (hunk 11 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Rename `size` to `out_shape` which is the shape of target output with format [N, C, H, W]

File: python/paddle/fluid/layers/nn.py (hunk 10 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Rename `size` to `out_shape` which is the shape of target output with format [N, C, H, W]

File: python/paddle/fluid/layers/nn.py (hunk 9 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/layers/nn.py]: Sorry. I forgot to remove this line.

File: python/paddle/fluid/tests/unittests/test_layers.py (hunk 8 above)
Comment (wanghaoshuang) [reply python/paddle/fluid/tests/unittests/test_layers.py:882]: I keep it because it is used by other test cases in test_layers.py

**Review by panyx0718 — DISMISSED** [review]
LG API

**Review by qingqing01 — APPROVED** [review]

**Review by typhoonzero — APPROVED** [review]
approve api

**Pull request merged**